src/kerberos.cc

```diff
@@ -75,128 +33,105 @@ NAN_METHOD(InitializeClient) {
         mech_oid = &spnego_mech_oid;
     }
 
-    AsyncQueueWorker(
-        new InitializeClientWorker(service, principal, gss_flags, mech_oid, callback));
-}
-
-class InitializeServerWorker : public Nan::AsyncWorker {
-   public:
-    InitializeServerWorker(std::string service, Nan::Callback* callback)
-        : AsyncWorker(callback, "kerberos:InitializeServerWorker"),
-          _service(service),
-          _server_state(NULL) {}
-
-    virtual void Execute() {
-        std::unique_ptr<gss_server_state, FreeDeleter> state(gss_server_state_new());
-        std::unique_ptr<gss_result, FreeDeleter> result(
-            authenticate_gss_server_init(_service.c_str(), state.get()));
+    KerberosWorker::Run(callback, "kerberos:InitializeClient", [=](KerberosWorker::SetOnFinishedHandler onFinished) {
+        gss_client_state* client_state = gss_client_state_new();
+        std::shared_ptr<gss_result> result(authenticate_gss_client_init(
+            service.c_str(), principal.c_str(), gss_flags, NULL, mech_oid, client_state), ResultDeleter);
 
+        // must clean up state if we won't be using it, smart pointers won't help here unfortunately
+        // because we can't `release` a shared pointer.
         if (result->code == AUTH_GSS_ERROR) {
-            SetErrorMessage(result->message);
-            return;
+            free(client_state);
```

Comment:
this is not making me happy, but I can't think of a solid way around it. Can we do something like

```c++
std::shared_ptr<gss_server_state> client_state;
std::shared_ptr<gss_result> result;
{
  std::shared_ptr<gss_server_state> _client_state(gss_client_state_new());
  result = std::make_shared<gss_result>(authenticate_gss_client_init(service.c_str(), principal.c_str(), gss_flags, NULL, mech_oid, client_state));
  if (result->code != AUTH_GSS_ERROR) {
    client_state = _client_state;
  }
}
```

Forgive my rusty c++ :)

Reply:
yeah unfortunately you can't `move` values with lambda capture until c++14 :(, I agree that just using `unique_ptr` like I had in the original code path would be ideal but we're only doing this in two spots so I think I can live with it..

Reply:
I could theoretically throw in a `scoped_ptr` implementation, but I think it's a bit egregious for two small cases